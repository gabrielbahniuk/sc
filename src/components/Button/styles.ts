import styled from 'styled-components'

export const Container = styled.button`
  width: 100%;
  padding: 0 1.5rem;
  height: 4rem;
  background: ${({ restart }): string => (restart ? '#f55359' : 'var(--secondary-color)')};
  color: #000;
  border-radius: 14px 7px 14px 7px;
  border: 0;
  font-size: 1.5rem;
  margin-top: 1.5rem;
  transition: filter 0.2s;
  font-weight: 600;
`
