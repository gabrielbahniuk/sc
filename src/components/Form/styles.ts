import styled from 'styled-components'

export const FormContainer = styled.form`
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 80%;

  label {
    font-size: 1.5rem;
    color: var(--text-color);
    width: 100%;
  }
  input,
  select {
    width: 100%;
    padding: 0 1.5rem;
    height: 4rem;
    border-radius: 0.25rem;
    border: 1px solid #d7d7d7;
    background: #e7e9ee;
    font-weight: 400;
    font-size: 1.5rem;
  }

  input[type='number']::-webkit-inner-spin-button,
  input[type='number']::-webkit-outer-spin-button {
    -webkit-appearance: none;
    -webkit-appearance: textfield;
    -moz-appearance: textfield;
    appearance: textfield;

    cursor: pointer;
    display: block;
    width: 1rem;
    color: #333;
    text-align: center;
    position: relative;
  }

  label + label {
    display: block;
    margin-top: 2rem;
  }
`
