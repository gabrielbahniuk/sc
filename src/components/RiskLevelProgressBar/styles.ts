import { appConfig } from '../../config'
import styled from 'styled-components'

const { MAXIMUM_RISK_LEVEL } = appConfig

export const ProgressBar = styled.div`
  background: #333;
  border-radius: 13px;
  height: 1.5rem;
  width: 100%;
  padding: 3px;
  transition: width 0.2s;
  margin-top: 1rem;

  span {
    content: '';
    display: block;
    background-color: ${({ riskLevel }): string =>
      riskLevel <= 8 ? 'var(--ok-color)' : riskLevel <= 16 ? 'var(--alert-color)' : 'var(--danger-color)'};
    width: ${({ riskLevel }): number => (riskLevel / MAXIMUM_RISK_LEVEL) * 100}%;
    height: 100%;
    border-radius: 9px;
    transition: width 0.3s;
  }
`

export const ProgressLabelContainer = styled.div`
  display: flex;
  width: 100%;
  justify-content: space-between;
  span {
    color: var(--text-color);
  }
`
