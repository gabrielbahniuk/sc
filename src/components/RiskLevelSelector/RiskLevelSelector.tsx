import { appConfig } from '@/config'
import React from 'react'

type RiskLevelSelector = {
  riskLevel: number
  handleRiskLevel(event: React.ChangeEvent<HTMLSelectElement>): void
}

const RiskLevelSelector: React.FC<RiskLevelSelector> = ({ riskLevel, handleRiskLevel }) => {
  const options = []
  const { MINIMUM_RISK_LEVEL, MAXIMUM_RISK_LEVEL } = appConfig

  for (let i = MINIMUM_RISK_LEVEL; i <= MAXIMUM_RISK_LEVEL; ++i) {
    options.push(
      <option key={i} value={i}>
        {i}
      </option>
    )
  }

  return (
    <select data-cy="riskLevelSelector" onChange={handleRiskLevel} value={riskLevel}>
      {options}
    </select>
  )
}

export { RiskLevelSelector }
