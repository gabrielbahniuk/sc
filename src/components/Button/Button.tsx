import React from 'react'

import { Container } from './styles'

type ButtonProps = {
  children?: React.ReactNode
  onClick?: any
  rest?: any
  [x: string]: any
}

const Button: React.FC<ButtonProps> = ({ onClick, children, ...rest }) => {
  return (
    <Container onClick={onClick} {...rest}>
      {children}
    </Container>
  )
}

export { Button }
