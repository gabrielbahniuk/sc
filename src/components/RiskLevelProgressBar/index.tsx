import React from 'react'

import { ProgressBar, ProgressLabelContainer } from './styles'

type RiskLevelProgressBar = {
  riskLevel: number
}

const RiskLevelProgressBar: React.FC<RiskLevelProgressBar> = ({ riskLevel }) => {
  return (
    <>
      <ProgressBar riskLevel={riskLevel}>
        <span data-cy="progressBar"></span>
      </ProgressBar>
      <ProgressLabelContainer>
        <span>Low Risk</span>
        <span>High Risk</span>
      </ProgressLabelContainer>
    </>
  )
}

export default RiskLevelProgressBar
