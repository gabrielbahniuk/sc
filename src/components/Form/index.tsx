import React from 'react'
import { useHistory } from 'react-router-dom'
import { useInvestments } from '@/hooks'
import RiskLevelSelector from '@/components/RiskLevelSelector'
import RiskLevelProgressBar from '@/components/RiskLevelProgressBar'
import Button from '@/components/Button'

const Form: React.FC = () => {
  const history = useHistory()
  const {
    initialInvestment,
    riskLevel,
    amountInvestmentYears,
    handleInitialInvestment,
    handleRiskLevel,
    handleAmountInvestmentYears,
    MINIMUM_INITIAL_INVESTMENT,
    MINIMUM_YEARS_INVESTMENT,
    MAXIMUM_YEARS_INVESTMENT
  } = useInvestments()

  function handleNextPage(): void {
    history.push('/result')
  }

  return (
    <form>
      <label>
        Initial Investment
        <input
          type="number"
          step="1000"
          data-cy="initialInvestment"
          min={MINIMUM_INITIAL_INVESTMENT}
          value={initialInvestment}
          onChange={handleInitialInvestment}
        />
      </label>
      <label>
        How many years?
        <input
          type="number"
          min={MINIMUM_YEARS_INVESTMENT}
          max={MAXIMUM_YEARS_INVESTMENT}
          step="1"
          data-cy="amountInvestmentYears"
          value={amountInvestmentYears}
          onChange={handleAmountInvestmentYears}
        />
      </label>
      <label>
        Risk Level
        <RiskLevelSelector data-cy="riskLevelSelector" riskLevel={riskLevel} handleRiskLevel={handleRiskLevel} />
      </label>
      <RiskLevelProgressBar riskLevel={riskLevel} />
      <Button data-cy="btnCalculate" onClick={handleNextPage}>
        Calculate
      </Button>
    </form>
  )
}

export default Form
