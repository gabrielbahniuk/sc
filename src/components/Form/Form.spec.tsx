import React from 'react'
import renderer from 'react-test-renderer'
import { shallow, ShallowWrapper } from 'enzyme'
import Button from '@/components/Button'
import RiskLevelSelector from '@/components/RiskLevelSelector'
import RiskLevelProgressBar from '@/components/RiskLevelProgressBar'
import Form from '.'

describe('Test Form component', () => {
  it('should render Form', () => {
    const form = renderer.create(<Form />).toJSON()
    expect(form).toMatchSnapshot()
  })
  it('should have correct elements within Form', () => {
    const wrapper: ShallowWrapper = shallow(<Form />)
    expect(wrapper.find('input').length).toBe(2)
    expect(wrapper.find('label').length).toBe(3)
    expect(wrapper.find(Button).length).toBe(1)
    expect(wrapper.find(RiskLevelSelector).length).toBe(1)
    expect(wrapper.find(RiskLevelProgressBar).length).toBe(1)
  })
})
