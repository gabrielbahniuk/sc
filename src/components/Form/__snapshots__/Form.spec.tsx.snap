// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Form component should render Form 1`] = `
<form
  className="sc-hKFxyN dsJelp"
>
  <label>
    Initial Investment
    <input
      data-cy="initialInvestment"
      step="1000"
      type="number"
    />
  </label>
  <label>
    How many years?
    <input
      data-cy="amountInvestmentYears"
      step="1"
      type="number"
    />
  </label>
  <label>
    Risk Level
    <select
      data-cy="riskLevelSelector"
    >
      <option
        value={3}
      >
        3
      </option>
      <option
        value={4}
      >
        4
      </option>
      <option
        value={5}
      >
        5
      </option>
      <option
        value={6}
      >
        6
      </option>
      <option
        value={7}
      >
        7
      </option>
      <option
        value={8}
      >
        8
      </option>
      <option
        value={9}
      >
        9
      </option>
      <option
        value={10}
      >
        10
      </option>
      <option
        value={11}
      >
        11
      </option>
      <option
        value={12}
      >
        12
      </option>
      <option
        value={13}
      >
        13
      </option>
      <option
        value={14}
      >
        14
      </option>
      <option
        value={15}
      >
        15
      </option>
      <option
        value={16}
      >
        16
      </option>
      <option
        value={17}
      >
        17
      </option>
      <option
        value={18}
      >
        18
      </option>
      <option
        value={19}
      >
        19
      </option>
      <option
        value={20}
      >
        20
      </option>
      <option
        value={21}
      >
        21
      </option>
      <option
        value={22}
      >
        22
      </option>
      <option
        value={23}
      >
        23
      </option>
      <option
        value={24}
      >
        24
      </option>
      <option
        value={25}
      >
        25
      </option>
    </select>
  </label>
  <div
    className="sc-gtsrHT cVxNMu"
  >
    <span
      data-cy="progressBar"
    />
  </div>
  <div
    className="sc-dlnjwi gvUrps"
  >
    <span>
      Low Risk
    </span>
    <span>
      High Risk
    </span>
  </div>
  <button
    className="sc-bdnxRM dmWuXT"
    data-cy="btnCalculate"
    onClick={[Function]}
    type="button"
  >
    Calculate
  </button>
</form>
`;
